src/gwt/src/org/rstudio/studio/client/common/spelling/RealtimeSpellChecker.java

```diff
@@ -0,0 +1,313 @@
+import org.rstudio.studio.client.common.spelling.model.SpellCheckerResult;
+import org.rstudio.studio.client.server.ServerError;
+import org.rstudio.studio.client.server.ServerRequestCallback;
+import org.rstudio.studio.client.workbench.WorkbenchList;
+import org.rstudio.studio.client.workbench.WorkbenchListManager;
+import org.rstudio.studio.client.workbench.events.ListChangedEvent;
+import org.rstudio.studio.client.workbench.prefs.model.UserPrefs;
+import org.rstudio.studio.client.workbench.views.source.editors.text.spelling.SpellingDoc;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+
+public class RealtimeSpellChecker
+{
+   public interface Context
+   {
+      ArrayList<String> readDictionary();
+      void writeDictionary(ArrayList<String> words);
+
+      void invalidateAllWords();
+      void invalidateWord(String word, boolean userDictionary);
+
+      void releaseOnDismiss(HandlerRegistration handler);
+   }
+
+   interface Resources extends ClientBundle
+   {
+      @Source("./domain_specific_words.csv")
+      TextResource domainSpecificWords();
+   }
+
+   public RealtimeSpellChecker(Context context)
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
+
+      // save reference to context and read its dictionary
+      context_ = context;
+      contextDictionary_ = context_.readDictionary();
+
+      // subscribe to spelling prefs changes (invalidateAll on changes)
+      ValueChangeHandler<Boolean> prefChangedHandler = (event) -> context_.invalidateAllWords();
+      ValueChangeHandler<Boolean> realtimeChangedHandler = (event) -> {};
+      ValueChangeHandler<String> dictChangedHandler = (event) -> {};
+      userPrefs_.ignoreUppercaseWords().addValueChangeHandler(prefChangedHandler);
+      userPrefs_.ignoreWordsWithNumbers().addValueChangeHandler(prefChangedHandler);
+      userPrefs_.spellingDictionaryLanguage().addValueChangeHandler(dictChangedHandler);
+      userPrefs_.realTimeSpellchecking().addValueChangeHandler(realtimeChangedHandler);
+
+      // subscribe to user dictionary changes
+      context_.releaseOnDismiss(userDictionary_.addListChangedHandler((ListChangedEvent event) ->
+         {
+            // detect whether this is the first delivery of the list
+            // or if it is an update
+            userDictionaryWords_ = event.getList();
+            updateIgnoredWordsIndex();
+         }
+      ));
+   }
+
+   @Inject
+   void initialize(SpellingService spellingService, WorkbenchListManager workbenchListManager, UserPrefs uiPrefs)
+   {
+      spellingService_ = spellingService;
+      userDictionary_ = workbenchListManager.getUserDictionaryList();
+      userPrefs_ = uiPrefs;
+
+      if (domainSpecificWords_.isEmpty())
+      {
+         String[] words = RES.domainSpecificWords().getText().split("[\\r\\n]+");
+         for (String w : words)
+         {
+            if (w.length() > 0)
+               domainSpecificWords_.add(w.toLowerCase());
```

Comment:
Is it necessary for us to convert to lowercase here? I'm thinking of words where capitalization might be important, e.g. for names like Bioconductor (https://www.bioconductor.org/)

Reply:
This is old untouched code (most of RealtimeSpellChecker is actually untouched, github didn't process the rename as I had hoped) but I think this is a fair change.